### PR TITLE
server: support command line output config info

### PR DESCRIPTION
**Pull request by glorv**
Signed-off-by: glorv <glorvs@163.com>

### What is changed and how it works?

Issue Number: ref #12492

What's Changed:

```
Add a new command line parameter `--config-info json` to output all config infos. Currently, the command only support Name, DefaultValue and ValueInFile 3 fields.
```

This is a simplified implementation of #12517 that only support output "Name, DefaultValue and ValueInFile", so we can avoid using the complex proc-macro.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Add a new command line parameter `--config-info` to output some information of all configs.
```

### CODE REVIEW

**Comment by glorv** [comment]
@BusyJay @tabokie PTAL, thank you

File: components/server/src/server.rs [hunk 1]
```diff
@@ -1851,3 +1854,118 @@ impl IOBudgetAdjustor for EnginesResourceInfo {
...
+
+    fn flatten_value(
+        default_obj: &Map<String, Value>,
+        value_obj: &Map<String, Value>,
+        key_buf: &mut String,
+        res: &mut Vec<Value>,
+    ) {
+        for (k, v) in default_obj.iter() {
+            let cfg_val = value_obj.get(k);
+            append_sub_key(key_buf, k);
```
Comment (tabokie) [inline components/server/src/server.rs:1907]: These tiny functions don't make it easy to review. How about
```suggestion
            let old_len = key_buf.len();
            if !k.is_empty() {
              key.push_str(k);
              key.push('.');
            }
```

File: components/server/src/server.rs [hunk 2]
```diff
@@ -1851,3 +1854,118 @@ impl IOBudgetAdjustor for EnginesResourceInfo {
...
+                flatten_value(
+                    v.as_object().unwrap(),
+                    cfg_val.unwrap().as_object().unwrap(),
+                    key_buf,
+                    res,
+                );
+            } else {
+                res.push(to_cfg_value(v, cfg_val, key_buf));
+            }
+            pop_sub_key(key_buf, k);
```
Comment (tabokie) [inline components/server/src/server.rs:1918]:
```suggestion
            key_buf.truncate(old_len);
```

File: components/server/src/server.rs [hunk 3]
```diff
@@ -1851,3 +1854,118 @@ impl IOBudgetAdjustor for EnginesResourceInfo {
         (total_budgets as f32 * score) as usize
     }
 }
+
+// currently, only support return json result
+pub fn get_flatten_cfg_info(cfg: &TiKvConfig) -> Vec<Value> {
```
Comment (tabokie) [inline components/server/src/server.rs:1859]: Put it near `TiKvConfig`?

**Comment by glorv** [comment]
@tabokie @BusyJay PTAL again, thanks~

**Review by BusyJay — APPROVED** [review]

**Review by tabokie — APPROVED** [review]
Could you paste a sample output? I'm not sure if deprecated items are printed.

**Comment by glorv** [comment]
> Could you paste a sample output? I'm not sure if deprecated items are printed.

emm.. The currently implement outputs all serializable config fields. The configs included are the same as all the configs in the `last_tikv.toml`. So deprecated items are also included. Do you have any good ideas to excludes them?

The current output is like:
```
[gl@ip-172-16-5-33 tikv]$ ./target/debug/tikv-server --config tests/integrations/config/test-custom.toml --config-info json
{
  "Component": "TiKV Server",
  "Version": "6.2.0-alpha",
  "Parameters": [
    {
      "Name": "log-level",
      "DefaultValue": "info"
    },
    {
      "Name": "log-file",
      "DefaultValue": ""
    },
    {
      "Name": "log-format",
      "DefaultValue": "text"
    },
    {
      "Name": "log-rotation-timespan",
      "DefaultValue": "0s",
      "ValueInFile": "1d"
    },
    {
      "Name": "log-rotation-size",
      "DefaultValue": "300MiB"
    },
    {
      "Name": "slow-log-file",
      "DefaultValue": "",
      "ValueInFile": "slow_foo"
    },
    {
      "Name": "slow-log-threshold",
      "DefaultValue": "1s"
    },
    {
      "Name": "panic-when-unexpected-key-or-data",
      "DefaultValue": false,
      "ValueInFile": true
    },
    {
      "Name": "abort-on-panic",
      "DefaultValue": false,
      "ValueInFile": true
    },
   ...
  ]
}
```

**Comment by tabokie** [comment]
@glorv deprecated items are all marked with `#[online_config(skip)]` and `#[serde(skip_serializing)]`. So I guess they won't appear in the list of `ConfigValue`?

You can verify it by checking a deprecated config e.g. `xxcf.titan.gc-merge-rewrite`.

**Comment by glorv** [comment]
> @glorv deprecated items are all marked with `#[online_config(skip)]` and `#[serde(skip_serializing)]`. So I guess they won't appear in the list of `ConfigValue`?
> 
> You can verify it by checking a deprecated config e.g. `xxcf.titan.gc-merge-rewrite`.
> 
> If `#[online_config(skip)]` doesn't work, maybe try `#[online_config(hidden)]`.

I have checked all the configs, there are only 8 commented with `// deprecated`. And there are 19 fields that are marked with `#[serde(skip_serializing)]`. These fields won't be included in the output.

**Comment by glorv** [comment]
@tabokie can we merge this pr now?

**Comment by tabokie** [comment]
Sure, I already approved it.

**Pull request merged**